### PR TITLE
Predict reconstruction on events without truth variables for loss and performance metrics

### PR DESCRIPTION
**Pull request by nickwp**
This PR allows the reconstruction engine to run a trained model over events without calculating loss or other performance metrics like classificaton accuracy or regression precision, etc.

Currently the whole WatChMaL code assumed that the true target values were available, so the changes to allow prediction without these values was much more involved than expected requiring many changes to both the reconstruction engine (and its subclasses) and the dataset class and configuration set up. Full details are below the summary.

**Summary of changes for users**
- The new `predict` method of the reconstruction class runs inference on data using a loaded trained model, like the `evaluate` method but without requiring any truth information and does not calculate loss or other metrics
  - HDF5 data files do not need to have any truth values (no labels for classification or target values for regression) when running predictions
  - The label set is not required in the classification config
  - The target key(s) are still required in the engine config even if loading and predicting from a state from a previous run, but this requirement could be removed in future
  - For regression, the saved state for prediction **must** be from a training run after this PR, otherwise the information on the target sizes (e.g. 3D for position, etc.) is not available, since the regression network itself only returns a concatenated array of values but it needs to know how to split this into each predicted quantity
    - As a workaround to use saved model states that were created before this PR, construct a configuration that loads the old state file and runs evaluate on some data as you would have done before this PR (requiring data that has truth info), but add another `save_state` to the end of the tasks config like you would in a training run; this will save a new state file that's a copy of the old one but with the extra info so it can later be used for prediction without truth info
- The loss function needs to be defined in both the `train` task and/or the `evaluate` task
  - This is because both of these calculate loss, while other tasks don't, but a future refactor should remove this duplication so that it only needs to be defined once for the engine
  - The default loss function for evaluation (same as train) is cross-entropy loss used for classification, so **regression configs should be updated to explicitly override the evaluate loss function** (like they should already be overriing the train loss function)

**Changes to reconstruction engine**
- The `evaluate` method of the reconstruction engine is now `inference`, which can run either with or without metrics calculations so it is wrapped by two methods:
  -  `evaluate` to act the same as it did before this PR, for evaluating data using a trained model, making predictions, and calculating the loss and reconstruction quality metrics
  - `predict` to perform only the reconstruction predictions on data using a trained model
- The `forward` method is split into `forward_pass` and `compute_metrics` and replaced by the `step` method that wraps these functions, always doing the forward pass of the model but then also computing metrics only where needed
- After loading the data loaders, the target key is passed to the dataset classes so it can load the relevant target values if they exist
- `process_data` is split into `process_data` (now defined in the base class as it's the same for classification and regression) and `process_target` defined in the subclasses
- **For the `ClassifierEngine`:**
  - Only changes are to implement the above changes to the base class
- **For the `RegressionEngine`:**
  - When saving and restoring a state, the information on target sizes (e.g. 3D for position reconstruction) is included. This allows loading a pre-trained model and predicting the target without any truth information (where prevously the truth information informed what the size of the target should be)
  - Slight modification to how the targets are stacked when processing the model output, which should improve compatibility for multiple regression targets of different sizes

**Changes to the `H5Dataset` class**
- Truth arrays that may not exist (labels, energies, positions, angles) are no longer automatically read into memory under the assumption they exist in the HDF5 file
- Target arrays required for training or evaluation are now read from the H5 file, if the file contains the true target array, into memory programmatically using `set_target`, which is called by the reconstrution engine after creating the data loaders (and does nothing if the target array does not exist, so that prediction can run without truth values)
- Dataset length is now defined by the length of the `event_hits_index` array (that always should exist) rather than the `labels` array (that does not need to exist)

**Changes to configs**
- The `evaluate` task sets its own loss function by default, like the `train` task config already does
  - This is a bit annoying because it means the loss function is defined twice, once for train and once for evaluate, if doing both, so a future refactor should probably move the loss config to the engine itself rather than the tasks
- A default/example `predict/test.yaml` task config is added, which is like `evaluate/test.yaml` but allows prediction without truth values
- Some cleanup of other broken default or example configs

**Other**
- Bug fix to only pin memory if using GPU, to avoid occasional errors when using CPU only

### CODE REVIEW

**Review by AndrewAtta — APPROVED** [review]
Looks good. Tested on several different datasets (PG and beam MC) and everything seems to function as expected

**Review by pdeperio — APPROVED** [review]

**Comment by PoinRec** [comment]
I’ve done a preliminary prediction run for the data of WCTE calibration run, and everything went smoothly at first glance. For a more detailed accuracy analysis I might need additional time, but the overall pipeline appears to be working as intended.

A couple of points from my side:

1. For prediction mode, the model still requires a “dummy” split list where test_idxs simply run from 0 to N–1, as Andrew has mentioned in the WatChMaL meeting.
2. For those who don’t use prediction mode, they will still need to slightly update their config files for regression training once the PR has been merged as is mentioned in Nick's comment.
